### DISCUSSION TITLE
Dynamic fetching of GTM

### DISCUSSION THREAD

**Discussion by ZeldOcarina** [gatsbyjs/gatsby#36380] [Help]
Hi!
I am trying to build a theme that I can reuse multiple times by fetching a CRM.
One of the components is injecting a GTM but I can't find a way of adding that dynamically (by fetching a promise to the CRM on gatsby-config).
Is there a way to achieve that?

**Comment by TylerBarnes** [comment]
@ZeldOcarina Gatsby themes are able to return a function that returns a config in gatsby-config.js. See [this example](https://github.com/gatsbyjs/themes/blob/master/packages/gatsby-theme-blog/gatsby-config.js). I'm not 100% sure if this function can return a promise or not, but if it can that will be the way to achieve what you're after. If it can't then I would recommend using env vars for that plugin option.

**Comment by ZeldOcarina** [comment]
Yeah that might be a solution but it requires a theme to function.
There's very little documentation on how to change the starter to a theme, does it have to be published to NPM?

**Reply by TylerBarnes** [reply]
You could make it a local plugin in your starter https://www.gatsbyjs.com/docs/creating-a-local-plugin/. It doesn't require being published to NPM

**Comment by ZeldOcarina** [comment]
Thank you!
I managed with this solution:
https://stackoverflow.com/questions/63508538/how-can-i-get-dynamic-data-from-within-gatsby-config-js
I'm not sure if it's the cleanest but it seems to be working for now! :)

**Reply by TylerBarnes** [reply]
Mutating plugin options could have unexpected consequences and cause weird bugs so I wouldn't recommend it, but perhaps it's harmless enough in this case

**Reply by ZeldOcarina** [reply]
Yeah I see your point..
I'll add this to my tech debt and call it a day for now haha
Thanks a lot for your help!

**Reply by TylerBarnes** [reply]
Anytime!

**Comment by ZeldOcarina** [comment]
Hi!
I've found a solution!
I've built my own Node app to prepare the .env file before the actual build so now there are two build processes lol.

I still think Gatsby should support async gatsby-config it would be so much easier.

Here's a gist example:
https://gist.github.com/ZeldOcarina/94b46dd61a11a2075c664a23fcccd9ba

Then I just run "NODE_ENV=development node create-env.js && gatsby develop"
